### PR TITLE
STY: Removed unconcatenated strings

### LINKED ISSUE #30454: CI: code check for " " introduced by black
Sometimes when we run `black` on something like

```
foo = (
    "bar "
    "baz"
)       
```

we end up with `foo = "bar " "baz"` when we would want `foo = "bar baz"`.

This is [not](https://github.com/psf/black/issues/1051) considered a bug in `black`, so is something we're currently checking for manually.  We should add it to code_checks.

### PR DESCRIPTION
**Pull request by ShaharNaveh**
- [x] ref #30454
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: scripts/validate_string_concatenation.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+"""
+Check where there is a string that needs to be concatenated.
+"""
+
+import os
+import sys
+import token
+import tokenize
+
+FILE_EXTENTIONS_TO_CHECK = [".py", ".pyx"]
```
Comment (MomIsBestFriend) [inline scripts/validate_string_concatenation.py]: What other file types should be included?

Comment (alimcmaster1) [reply scripts/validate_string_concatenation.py]: Maybe .pxd and and .pxi.ini?

File: scripts/validate_string_concatenation.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    Checking if the file containing strings that needs to be concatenated.
+
+    Parameters
+    ----------
+    file_path : str
+        File path pointing to a single file.
+
+    Returns
+    -------
+    int
```
Comment (alimcmaster1) [inline scripts/validate_string_concatenation.py]: Feels like a bool would be a more appropriate return type if this function is only ever going to return 0/1

File: scripts/validate_string_concatenation.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            if ext in FILE_EXTENTIONS_TO_CHECK:
+                status_codes.add(is_concatenated(os.path.join(subdir, file_name)))
+
+    if 1 in status_codes:
+        sys.exit(1)
+
+    sys.exit(0)
+
+
+def is_concatenated(file_path):
```
Comment (alimcmaster1) [inline scripts/validate_string_concatenation.py]: Could add type?

File: scripts/validate_string_concatenation.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        1 - Needs to be fixed.
+    """
+    with open(file_path, "r") as file_name:
+        toks = list(tokenize.generate_tokens(file_name.readline))
+        for i in range(len(toks) - 1):
+            tok = toks[i]
+            tok2 = toks[i + 1]
+            if tok[0] == token.STRING and tok[0] == tok2[0]:
+                print(
+                    "{file_path}:{line_number}:\t{start} and {end}".format(
```
Comment (alimcmaster1) [inline scripts/validate_string_concatenation.py]: f string?

File: scripts/validate_string_concatenation.py (hunk 2 above)
Comment (MomIsBestFriend) [reply scripts/validate_string_concatenation.py]: I did also thought so, but I am relying on an exit status, so int feels more native.

File: scripts/validate_string_concatenation.py (hunk 3 above)
Comment (MomIsBestFriend) [reply scripts/validate_string_concatenation.py]: At first I typed this whole script, but it said it was a syntax error, same with ```f-strings```, I think that the running platform of this is python 3.5

File: scripts/validate_string_concatenation.py (hunk 4 above)
Comment (MomIsBestFriend) [reply scripts/validate_string_concatenation.py]: Oh, I already answered above :smile:

File: scripts/validate_string_concatenation.py (hunk 3 above)
Comment (alimcmaster1) [reply scripts/validate_string_concatenation.py]: Hmm - we run `validate_docstrings.py` as part of `code_checks.sh` which has 3.6 + syntax?

File: ci/code_checks.sh [hunk 5]
```diff
@@ -207,6 +207,10 @@ if [[ -z "$CHECK" || "$CHECK" == "patterns" ]]; then
     invgrep -R --include=*.{py,pyx} 'xrange' pandas
     RET=$(($RET + $?)) ; echo $MSG "DONE"
 
+    MSG='Check for use of not concatenated strings' ; echo $MSG
+    python ./scripts/validate_string_concatenation.py pandas
```
Comment (alimcmaster1) [inline ci/code_checks.sh]: Should this use $BASE_DIR ?

Comment (MomIsBestFriend) [reply ci/code_checks.sh]: Correct, will fix

File: scripts/validate_string_concatenation.py (hunk 3 above)
Comment (MomIsBestFriend) [reply scripts/validate_string_concatenation.py]: > Hmm - we run `validate_docstrings.py` as part of `code_checks.sh` which has 3.6 + syntax?

I really don't know.

File: scripts/validate_string_concatenation.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python
+"""
+Check where there is a string that needs to be concatenated.
+"""
```
Comment (jbrockmendel) [inline scripts/validate_string_concatenation.py]: can you flesh this out with a GH reference and a sentence about how `black` causes the issue

File: scripts/validate_string_concatenation.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+import os
+import sys
+import token
+import tokenize
+
+FILE_EXTENTIONS_TO_CHECK = [".pxd", ".py", ".pyx", ".pyx.ini"]
```
Comment (jbrockmendel) [inline scripts/validate_string_concatenation.py]: EXTENTIONS -> EXTENSIONS

Comment (MomIsBestFriend) [reply scripts/validate_string_concatenation.py]: I need grammar lessons ASAP.

**Comment by datapythonista** [comment]
@MomIsBestFriend do you mind opening a PR for the script and the call to the script in `ci/code_checks.sh`? And just leave here the fixes.

The fixes should be non-trivial to get merged, but not the script (I think we should try using regex for it). Thanks!

**Comment by MomIsBestFriend** [comment]
> @MomIsBestFriend do you mind opening a PR for the script and the call to the script in `ci/code_checks.sh`? And just leave here the fixes.
> 
> The fixes should be non-trivial to get merged, but not the script (I think we should try using regex for it). Thanks!

Sure

**Review by datapythonista — APPROVED** [review]
Thanks @MomIsBestFriend, really nice clean up.

**Comment by jreback** [comment]
thanks @MomIsBestFriend

File: scripts/validate_string_concatenation.py (hunk 3 above)
Comment (MomIsBestFriend) [reply scripts/validate_string_concatenation.py]: > Hmm - we run `validate_docstrings.py` as part of `code_checks.sh` which has 3.6 + syntax?

@alimcmaster1 https://github.com/pandas-dev/pandas/pull/30467#issuecomment-569092377

**Pull request merged**